### PR TITLE
fix(config/validation): improve validation of global options

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
Add validation for globalOptions.

This PR changes the behaviour of `config-validator` when in comes to validating the file config. Till now we couldn't validate the global options inside file config fully ... even after adding logic to warn for any global option in repo config. We used to get `invalid config option error` instead if a proper validation erro... This has been fixed and now we can valiate all the config options, based on their types, and allowedValues (where needed). 
> This doesn't effect the existing behaviour of repo config validation. 

I decided to create a new function to validate the global options even though it was possible to modify the existing function for our use case, because we want to warn for any validation errors and not throw errors. Since the existing function is designed mainly for errors adding warnings for the global options would've made it less redable and more nested.

## Context
Needed For: https://github.com/renovatebot/renovate/pull/26379

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by rarkins — COMMENTED** [review]
Please expand the description as this is too generic/non-descriptive. What is this PR changing and why?

File: lib/config/validation.ts
Comment (HonkingGoose) [inline lib/config/validation.ts:770]:
```suggestion
        message: `Configuration option \`${currentPath}\` should be a JSON object`,
```

File: lib/config/validation.ts
Comment (HonkingGoose) [inline lib/config/validation.ts:746]:
```suggestion
        message: `Configuration option \`${currentPath}\` should be a boolean. Found: ${JSON.stringify(
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/config/validation.spec.ts
Comment (HonkingGoose) [inline lib/config/validation.spec.ts:918]: @rarkins you're a native speaker, which is the correct format?

1. validates options of string type
2. validates options of type string
3. validates string type options
4. validates options with type string

I suspect `3` is the correct format, but I'm asking you first to make sure. 😄

File: lib/config-validator.ts
Comment (rarkins) [inline lib/config-validator.ts:24]: careful, when you end in two booleans then it could be easy to miss some migrations here. Look at line 131 for example

File: lib/config/validation.spec.ts
Comment (rarkins) [reply lib/config/validation.spec.ts:918]: All are ok for me for a test description

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
What is the expected difference before/after this PR?

I just ran with an invalid option in config.js and did not get any warning or error

**Comment by rarkins** [comment]
The log:

```
❯ pnpm start --dry-run=extract renovate-reproductions/24460

> renovate@0.0.0-semantic-release prestart /Users/rhys/src/github/renovatebot/renovate
> run-s 'generate:*'


> renovate@0.0.0-semantic-release generate:imports /Users/rhys/src/github/renovatebot/renovate
> node tools/generate-imports.mjs

generating imports
> data/debian-distro-info.json
> data/kubernetes-api.json5
> data/node-js-schedule.json
> data/ubuntu-distro-info.json
> node_modules/emojibase-data/en/shortcodes/github.json
generating hashes

> renovate@0.0.0-semantic-release start /Users/rhys/src/github/renovatebot/renovate
> ts-node lib/renovate.ts "--dry-run=extract" "renovate-reproductions/24460"

DEBUG: Using RE2 as regex engine
DEBUG: Parsing configs
DEBUG: Checking for config file in /Users/rhys/src/github/renovatebot/renovate/config.js
DEBUG: Enabling debug logging to tmp/logs/2023-10-17T11:30:39.291Z.log
DEBUG: File config
       "config": {
         "platform": "github",
         "token": "***********",
         "binarySource": "docker",
         "baseDir": "/tmp/renovate",
         "onboardingConfig": {"extends": ["config:recommended"]},
         "prHourlyLimit": 0,
         "prConcurrentLimit": 0,
         "repositories": ["renovate-reproductions/24460"],
         "force": {"schedule": null},
         "logFile": "tmp/logs/2023-10-17T11:30:39.291Z.log",
         "allowCustomCrateRegistries": "nope"
       }
DEBUG: CLI config
       "config": {"repositories": ["renovate-reproductions/24460"], "dryRun": "extract"}
DEBUG: Env config
       "config": {"hostRules": []}
DEBUG: Combined config
       "config": {
         "platform": "github",
         "token": "***********",
         "binarySource": "docker",
         "baseDir": "/tmp/renovate",
         "onboardingConfig": {"extends": ["config:recommended"]},
         "prHourlyLimit": 0,
         "prConcurrentLimit": 0,
         "repositories": ["renovate-reproductions/24460"],
         "force": {"schedule": null},
         "logFile": "tmp/logs/2023-10-17T11:30:39.291Z.log",
         "allowCustomCrateRegistries": "nope",
         "hostRules": [],
         "schedule": null,
         "dryRun": "extract"
       }
DEBUG: Enabling forkProcessing while in non-autodiscover mode
DEBUG: Found valid git version: 2.38.1
DEBUG: Using default github endpoint: https://api.github.com/
DEBUG: Platform config
       "platformConfig": {
         "hostType": "github",
         "endpoint": "https://api.github.com/",
         "isGHApp": false,
         "isGhe": false,
         "userDetails": {
           "username": "renovate-tests",
           "name": "Renovate Testing",
           "id": 29705663
         },
         "userEmail": "renovate-tests@redgc.com"
       },
       "renovateUsername": "renovate-tests"
DEBUG: Using platform gitAuthor: Renovate Testing <renovate-tests@redgc.com>
```

My expectation was that the `"allowCustomCrateRegistries": "nope"` field would trigger a warning.

**Comment by RahulGautamSingh** [comment]
> What is the expected difference before/after this PR?
> 
> I just ran with an invalid option in config.js and did not get any warning or error

This is because we do not validate file config in repo runs.
To catch the errors you need to run `config-validator`

File: lib/config-validator.ts
Comment (viceice) [reply lib/config-validator.ts:24]:
```suggestion
  isGlobalConfig: boolean,
  isPreset: boolean
```

force parameter to not get into migration trouble

File: lib/config-validator.ts
Comment (viceice) [inline lib/config-validator.ts:84]:
```suggestion
          await validate(file, parsedContent, strict, false, false);
```

File: lib/config-validator.ts
Comment (viceice) [inline lib/config-validator.ts:105]:
```suggestion
          await validate(file, parsedContent, strict, false, false);
```

File: lib/config-validator.ts
Comment (viceice) [inline lib/config-validator.ts:125]:
```suggestion
          false,
          false,
```

File: lib/config-validator.ts
Comment (viceice) [inline lib/config-validator.ts:148]:
```suggestion
          await validate(file, fileConfig, strict, true, false);
```

File: lib/config/validation.spec.ts
Comment (viceice) [reply lib/config/validation.spec.ts:918]: i like 3

File: lib/config/validation.ts
Comment (viceice) [inline lib/config/validation.ts:102]:
```suggestion
  isGlobalConfig: boolean
```

make mandatory to not messup migration

**Review by viceice — COMMENTED** [review]
we should get this done

**Comment by RahulGautamSingh** [comment]
> we should get this done

Blocked by: https://github.com/renovatebot/renovate/pull/24561

File: lib/config/validation.ts
Comment (viceice) [inline lib/config/validation.ts:33]: 🤔 ` optionsGlobals` sounds strange. should we use `globalOptions` instead?

@rarkins @secustor WDYT?

File: lib/config/validation.ts
Comment (viceice) [inline lib/config/validation.ts:122]:
```suggestion
    optionGlobals = new Set();
```
does typescript auto detect type?

File: lib/config/validation.ts
Comment (viceice) [inline lib/config/validation.ts:751]: use tsdoc comment style.

**Review by viceice — COMMENTED** [review]
we should refactor validation to use same style as config migration for easier future additions and easier testing.

File: lib/config-validator.ts
Comment (RahulGautamSingh) [inline lib/config-validator.ts:131]: Revert

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:284]:
```suggestion
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:471]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:499]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:528]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:565]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:596]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:618]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:639]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:663]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:686]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:721]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:734]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:747]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:760]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:893]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:908]:
```suggestion
        true,
```

File: lib/config/validation.spec.ts
Comment (RahulGautamSingh) [inline lib/config/validation.spec.ts:947]:
```suggestion
        true,
```

**Review by RahulGautamSingh — COMMENTED** [review]
Revert

**Comment by RahulGautamSingh** [comment]
Redoing this: I think we actually do not a separate function we can add the validations inside the existing function.

File: lib/config/validation.ts
Comment (HonkingGoose) [inline lib/config/validation.ts]: I think we're missing a final `.` at the end of the message?

File: lib/config/validation.ts
Comment (HonkingGoose) [inline lib/config/validation.ts]: Same thing about the `.` here.

File: lib/config/validation.ts
Comment (HonkingGoose) [inline lib/config/validation.ts]: Same thing about the `.` here.

File: lib/config/validation.ts
Comment (HonkingGoose) [inline lib/config/validation.ts]: Same thing about the `.` here.

File: lib/config/validation.ts
Comment (HonkingGoose) [inline lib/config/validation.ts]: Same thing about the `.` here.

File: lib/config/validation.ts
Comment (HonkingGoose) [inline lib/config/validation.ts]: Same thing about the `.` here.

File: lib/config/validation.ts
Comment (HonkingGoose) [inline lib/config/validation.ts]: Same thing about the `.` here.

File: lib/config/validation.ts
Comment (HonkingGoose) [inline lib/config/validation.ts:889]:
```suggestion
              message: `Invalid \`${currentPath}.${subKey}\` configuration: value is not an integer.`,
```

File: lib/config/validation.ts
Comment (HonkingGoose) [inline lib/config/validation.ts:898]:
```suggestion
            message: `Invalid \`${currentPath}.${key}.${res}\` configuration: value is not a string.`,
```

File: lib/config/validation.spec.ts
Comment (HonkingGoose) [inline lib/config/validation.spec.ts:33]:
```suggestion
    it('only warns for actual globals in repo config', async () => {
```

Rewriting the test name to avoid the double negative. Please check if this is correct. :wink:

File: lib/config/validation.ts
Comment (HonkingGoose) [inline lib/config/validation.ts:781]: I don't know about you maintainers, but the code for the `validateGlobalConfig` function is hard to read for me, because of all the `if` and `else if` statements.

Maybe you can extract some blocks of code into separate validation functions?

File: lib/config/validation.spec.ts
Comment (viceice) [inline lib/config/validation.spec.ts:1184]: use ts-expect-error comment instead of never

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-9.html#-ts-expect-error-comments

File: lib/config/validation.ts
Comment (viceice) [inline lib/config/validation.ts:38]: can you please do this rename on a separate PR, so we've less changes here?

File: lib/config/validation.ts
Comment (viceice) [reply lib/config/validation.ts:781]: yes, I think this should be modelled like the migration service, so we've small validation functions which can be easily tested.

File: lib/config/validation.spec.ts
Comment (viceice) [inline lib/config/validation.spec.ts:32]: please do code move in separate PR for easier review

File: lib/config/validation.ts
Comment (RahulGautamSingh) [reply lib/config/validation.ts:781]: I think this should be done seperately under its own issue, will create one

File: lib/config/validation.spec.ts
Comment (HonkingGoose) [inline lib/config/validation.spec.ts:982]:
```suggestion
          message: `The "customEnvVariables" option is a global option reserved only for Renovate's global configuration and cannot be configured within repository config file,`,
```

I think we should say `Renovate's` here, instead of `bot's`.

Also a sentence should end with a `.`. :wink:

File: lib/config/validation.spec.ts
Comment (HonkingGoose) [inline lib/config/validation.spec.ts:1263]:
```suggestion
            'Invalid `cacheTtlOverride.someField` configuration: value must be an integer.',
```

It seems nice to say what the expected/valid input is, instead of the thing that's broken. I'll let the maintainers decide what's best. :wink:

File: lib/config/validation.spec.ts
Comment (HonkingGoose) [inline lib/config/validation.spec.ts:1279]:
```suggestion
            'Invalid `customEnvVariables.customEnvVariables.example2` configuration: value must be a string.',
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/config/validation.ts
Comment (viceice) [inline lib/config/validation.ts:153]: this skips validation for non-global options inside global config file. they shouldn't skipped

File: lib/config/validation.ts
Comment (viceice) [inline lib/config/validation.ts:152]:
```suggestion
    if (isGlobalConfig && validateGlobalConfig(key, val, optionTypes[key], warnings, currentPath)) {
```

Idea, `validateGlobalConfig` should return true when it validated the `key` and false when not, so it can fall through below to validate other options

or only call it when `key` is a global option

File: lib/config/validation.ts
Comment (viceice) [inline lib/config/validation.ts:908]:
```suggestion
  }
  return false;
```

**Review by viceice — CHANGES_REQUESTED** [review]

**Review by viceice — APPROVED** [review]

**Review by rarkins — COMMENTED** [review]
Is this a feature or a fix? There's seems to have been some logic here already which is improved/fixed

**Comment by RahulGautamSingh** [comment]
> Is this a feature or a fix? There's seems to have been some logic here already which is improved/fixed

Should be a fix, but the issue is marked as a feature hence I added the feature scope.

**Review by rarkins — APPROVED** [review]

**Comment by MindTooth** [comment]
Not to bump this, but this kinda broke stuff, and I'm not complete sure how to fix this.  E.g.:

```
     "warnings": [
         {
           "topic": "Configuration Error",
           "message": "Configuration option `hostRules[0].token` should be a string."
         },
         {
           "topic": "Configuration Error",
           "message": "Invalid `onboardingConfig.onboardingConfig.extends` configuration: value must be a string."
         }
       ]
```

---

Taken the example from https://docs.renovatebot.com/examples/self-hosting/#usage even produces an error:

```
 WARN: Found errors in configuration
       "file": "config.js",
       "warnings": [
         {
           "topic": "Configuration Error",
           "message": "Invalid `onboardingConfig.onboardingConfig.extends` configuration: value must be a string."
         }
       ]
```

**Comment by RahulGautamSingh** [comment]
Not sure about the `hostRules` error, I couldn't reproduce it. Can you share you config for it?

The `onboardingConfig` one was reproduced, I will add a fix for it asap.

**Comment by MindTooth** [comment]
```js
  hostRules: [
    {
      matchHost: "gitlab.<the rest>",
      token: process.env.RENOVATE_TOKEN,
    },
  ],
```

That might however be my own error? 🤔

**Comment by rarkins** [comment]
Is this from when you run it with the validator?

**Comment by MindTooth** [comment]
> Is this from when you run it with the validator?

Correct.  And it has now broken the pipeline, meaning the template from GitLab won't run.

**Comment by rarkins** [comment]
I think I can reproduce one of these. If you have a config.js, and it has a host rules entry with token pointing to an env variable, and that env variable doesn't exist when you run the validator, then the token=undefined and the validator will error. If instead the env variable is defined/valid then you don't get an error.

**Comment by rarkins** [comment]
@MindTooth please create a new github discussion to continue this topic

**Comment by HonkingGoose** [comment]
Do we need to fix the [documentation](https://docs.renovatebot.com/examples/self-hosting/#usage) or the code for the example:

> ## Usage
> 
> The following example uses the Renovate CLI tool, which you can install by running `npm i -g renovate`.
> 
> If running your own Renovate bot then you will need a user account that Renovate will run as.
> We recommend you create and use a dedicated account for the bot, e.g. name it `renovate-bot` if on your own instance.
> Create and save a PAT for this account.
> 
> Create a Renovate config file, for example:
> 
> ```js
> module.exports = {
>   endpoint: 'https://self-hosted.gitlab/api/v4/',
>   token: '**gitlab_token**',
>   platform: 'gitlab',
>   onboardingConfig: {
>     extends: ['config:recommended'],
>   },
>   repositories: ['username/repo', 'orgname/repo'],
> };
> ```
> 
> Here change the `logFile` and `repositories` to something appropriate.
> Also replace `gitlab-token` value with the one created during the previous step.
> 
> If you're running against GitHub Enterprise Server, then change the `gitlab` values in the example to the equivalent GitHub ones.
> 
> You can save this file as anything you want and then use the `RENOVATE_CONFIG_FILE` environment variable to tell Renovate where to find it.
> 
> Most people use `cron` to schedule when Renovate runs, usually on an hourly schedule.
> 
> ```sh title="Example bash script that you can point cron to"
> #!/bin/bash
> 
> export PATH="/home/user/.yarn/bin:/usr/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:$PATH"
> export RENOVATE_CONFIG_FILE="/home/user/renovate-config.js"
> export RENOVATE_TOKEN="**some-token**" # GitHub, GitLab, Azure DevOps
> export GITHUB_COM_TOKEN="**github-token**" # Delete this if using github.com
> 
> # Renovate
> renovate
> ```
> 
> Save the script file, and run the script manually.
> Only add the script to `cron` after you checked it works.
> 
> 
> !!! note
>     The GitHub.com token as an environment variable is needed to fetch changelogs that are usually hosted on github.com.
>     You don't need to add it if you are already running the bot against github.com, but you do need to add it if you're using GitHub Enterprise Server, GitLab, Azure DevOps, or Bitbucket.

**Comment by MindTooth** [comment]
> @MindTooth please create a new github discussion to continue this topic

https://github.com/renovatebot/renovate/discussions/27468

**Pull request merged**
